src/libtiled/mapobject.cpp

```diff
@@ -240,4 +230,83 @@ MapObject *MapObject::clone() const
     return o;
 }
 
+void MapObject::flipRectObject(FlipDirection direction, const QPointF &origin)
+{
+    QTransform flipTransform;
+    flipTransform.translate(origin.x(), origin.y());
+    qreal newRotation = 0;
+    if (direction == FlipHorizontally) {
+        newRotation = 180.0 - rotation();
+        flipTransform.scale(-1, 1);
+    } else { //direction == FlipVertically
+        flipTransform.scale(1, -1);
+        newRotation = -rotation();
+    }
+    flipTransform.translate(-origin.x(), -origin.y());
```

Comment:
The whole bit about determining the new rotation and the flip transform seems to be the same for each type of object. Please move that bit into `MapObject::flip` and pass the transform into the other functions instead of the direction and origin.